File: backend/src/main/java/com/loopmaster/backend/model/Cancion.java
package com.loopmaster.backend.model;

import jakarta.persistence.*;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "cancion")
public class Cancion {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @Column(name = "titulo", nullable = false, length = 200)
    private String titulo;

    @Column(name = "artista", length = 200)
    private String artista;

    @Column(name = "duracion_segundos", nullable = false)
    private int duracionSegundos;

    @Column(name = "nombre_archivo", length = 255)
    private String nombreArchivo;

    @OneToMany(mappedBy = "cancion", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Bucle> bucles = new ArrayList<>();

    @ManyToMany(mappedBy = "canciones")
    private Set<Carpeta> carpetas = new HashSet<>();

    public Cancion() {
        this.bucles = new ArrayList<>();
        this.carpetas = new HashSet<>();
    }

    public Cancion(int id, String titulo, String artista, int duracionSegundos) {
        this.id = id;
        this.titulo = titulo;
        this.artista = artista;
        this.duracionSegundos = duracionSegundos;
        this.bucles = new ArrayList<>();
        this.carpetas = new HashSet<>();
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getTitulo() {
        return titulo;
    }

    public void setTitulo(String titulo) {
        this.titulo = titulo;
    }

    public String getArtista() {
        return artista;
    }

    public void setArtista(String artista) {
        this.artista = artista;
    }

    public int getDuracionSegundos() {
        return duracionSegundos;
    }

    public void setDuracionSegundos(int duracionSegundos) {
        this.duracionSegundos = duracionSegundos;
    }

    public String getNombreArchivo() {
        return nombreArchivo;
    }

    public void setNombreArchivo(String nombreArchivo) {
        this.nombreArchivo = nombreArchivo;
    }

    public List<Bucle> getBucles() {
        return bucles;
    }

    public void setBucles(List<Bucle> bucles) {
        this.bucles = bucles;
    }

    public Set<Carpeta> getCarpetas() {
        return carpetas;
    }

    public void setCarpetas(Set<Carpeta> carpetas) {
        this.carpetas = carpetas;
    }

    @Override
    public String toString() {
        return "Cancion{" +
                "id=" + id +
                ", titulo='" + titulo + '\'' +
                ", artista='" + artista + '\'' +
                ", duracionSegundos=" + duracionSegundos +
                ", nombreArchivo='" + nombreArchivo + '\'' +
                '}';
    }
}

File: ios/LoopMaster/LoopMaster/Modelo/Cancion.swift
import Foundation
import SwiftData

@Model
final class Cancion {
    var titulo: String
    var artista: String
    var duracionSegundos: Int
    var nombreArchivo: String
    var fechaCreacion: Date

    var idServidor: Int?
    var pendienteSync: Bool = false
    var pendienteBorrado: Bool = false

    @Relationship(deleteRule: .cascade, inverse: \Bucle.cancion)
    var bucles: [Bucle] = []

    var carpetas: [Carpeta] = []

    init(
        titulo: String,
        artista: String = "",
        duracionSegundos: Int,
        nombreArchivo: String = "",
        idServidor: Int? = nil,
        pendienteSync: Bool = false,
        pendienteBorrado: Bool = false
    ) {
        self.titulo = titulo
        self.artista = artista
        self.duracionSegundos = duracionSegundos
        self.nombreArchivo = nombreArchivo
        self.fechaCreacion = Date()
        self.idServidor = idServidor
        self.pendienteSync = pendienteSync
        self.pendienteBorrado = pendienteBorrado
    }
}
